### PR TITLE
Add assertion IDs that will be included in production log statements for fail and hardAsserts

### PR DESCRIPTION
**Pull request by MarkDuckworth**
In the Firestore SDK, error messages are removed from `fail(message)` and `hardAssert(condition, message)` statements during the build. Removal of error messages is performed to reduce bundle size, however this comes at a cost to usability. Specifically, this means that production error messages can be unhelpful.

Here's an example of a production error message today (without this change):
```
FIRESTORE (1.2.3) INTERNAL ASSERTION FAILED: Unexpected state
```

"Unexpected state" is the same reason given for 130+ error conditions after the detailed error messages have been stripped.

To improve debuggability of error messages, this PR adds a numeric assertion ID to each `fail` and `hardAssert` statement. These numeric IDs are unique and will not be removed from the build output. The full error message will still be removed during the build, but the assertion IDs introduced with this change will give customers and support more information to debug the output.

Here's and example of a production error message after this change is applied:
```
FIRESTORE (1.2.3) INTERNAL ASSERTION FAILED: Unexpected state (ID: 0x2e4a) CONTEXT: {"example": "relevant JSON serialized state of the application when the assertion failed."}
```

Scripts:
`yarn assertion-id:new` - Create a new unique ID
`yarn assertion-id:check` - Check for duplicate and invalid id formatting
`yarn assertion-id:find <ID>` - Find an existing ID and the error message associated with it
`yarn assertion-id:list` - list all of the existing IDs and their error messages

### CODE REVIEW

**Comment by dconeybe** [comment]
IMO adding 2.4 kB to the bundle size is questionable if it's worth it.

**Comment by MarkDuckworth** [comment]
> IMO adding 2.4 kB to the bundle size is questionable if it's worth it.

@dconeybe, it's <=1% increase for making production issues more debuggable. In your opinion, what increase in bundle size would be acceptable?

**Comment by dconeybe** [comment]
> > IMO adding 2.4 kB to the bundle size is questionable if it's worth it.
> 
> @dconeybe, in your opinion, what increase in bundle size would be acceptable?
> 
> It's <=1% increase, which seems like a reasonable tradeoff for making production issues more debuggable. But I'm seeing if I can make the impact smaller.

I'd hope it would be less than 1kB. But this is just one man's opinion. Please don't take it as gospel truth.

**Review by dconeybe — APPROVED** [review]
This is fantastic! I especially like how you added the "context" which will be incredibly useful when debugging. Thanks for doing this, Mark.

**Comment by dconeybe** [comment]
Please let the team know about the 1% increase in code bundle size and make sure they're ok with it. The "firestore (Read Write w Persistence)" bundle increased by 2.77 kB which is probably worth it, but I'd make sure the team knows.

Also, please add a change set with a minor version bump and a change log entry.

**Comment by dconeybe** [comment]
FWIW I much prefer the "hashes" you had before over error codes numerically incrementing from zero. The reason is that hashes have much better "SEO" because of their uniqueness. Firstly, the improved SEO allows customers to search for the error code and find relevant bug reports, stack overflow posts, etc. If they search for error code "3" they won't likely get anything useful, but if they search for "jszxcbe37k" they'll likely get results that relate directly to the error that they are experiencing. Likewise, when we as Firestore SDK developers get a bug report with error code "3" it's much more difficult to find its place in the source code than searching for "jszxcbe37k".

For Data Connect I chose this alphabet: `[23456789abcdefghjkmnopqrstuvwxyz](https://github.com/firebase/firebase-android-sdk/blob/a0a02a73f67f6447c739d920f614c8bd5331c591/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/util/AlphanumericStringUtil.kt#L31C45-L31C79)` (a length of 32). This alphabet is most of the English lowercase letters and most digits, with those that "look alike" removed (e.g. zero "0" and lowercase-o "o"). This leads to incredibly unique 6-digit strings.

**Review by hsubox76 — APPROVED** [review]

File: .changeset/odd-wolves-sit.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@firebase/firestore": patch
+---
+
+Add assertion IDs and optional context objects that will be included in production log statements for fail and hardAsserts
```
Comment (dconeybe) [inline .changeset/odd-wolves-sit.md:5]: Consider a reword like this: Add unique IDs and state information into fatal error messages instead of the generic "unexpected state" message.

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        tsFiles = tsFiles.concat(getTsFilesRecursive(fullPath));
+      } else if (entry.isFile() && (entry.name.endsWith(".ts"))) {
+        // Exclude declaration files (.d.ts) as they usually don't contain implementation
+        if (!entry.name.endsWith(".d.ts")) {
+          tsFiles.push(fullPath);
+        }
+      }
+    }
+  } catch (error: any) {
+    console.error(`Error reading directory ${dirPath}: ${error.message}`);
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:80]: Consider simply letting the exception bubble up, possibly adding some contextual information. This error message could be easy to miss.

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+  let tsFiles: string[] = [];
+  try {
+    const entries = fs.readdirSync(dirPath, { withFileTypes: true });
+
+    for (const entry of entries) {
+      const fullPath = path.join(dirPath, entry.name);
+
+      if (entry.isDirectory()) {
+        // Ignore node_modules for performance and relevance
+        if (entry.name === 'node_modules') {
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:66]: Consider using the `git check-ignore` command to exclude any files that are ignored by git, rather than hardcoding specific directories. If you like this idea, feel free to implement this in a follow-up PR.

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+/**
+ * Recursively finds all files with .ts extensions in a directory.
+ * @param dirPath The absolute path to the directory to scan.
+ * @returns An array of absolute paths to the found TypeScript files.
+ */
+function getTsFilesRecursive(dirPath: string): string[] {
+  let tsFiles: string[] = [];
+  try {
+    const entries = fs.readdirSync(dirPath, { withFileTypes: true });
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:59]: Consider using a third-party library like "fast-glob" (https://www.npmjs.com/package/fast-glob) rather than implementing your own bespoke recursive directory search algorithm. Feel free to do this in a follow-up PR, if desired.

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+        // Continue traversing down the AST
+        ts.forEachChild(node, visit);
+      };
+
+      // Start traversal from the root SourceFile node
+      visit(sourceFile);
+
+    } catch (error: any) {
+      console.error(`Error processing file ${filePath}: ${error.message}`);
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:168]: Again, strongly consider just letting the exception bubble up and terminate the app. This "should" never fail, and if it does, IMO it should fail loudly.

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+  for (const filePath of filePaths) {
+    try {
+      // Read the file content
+      const sourceText = fs.readFileSync(filePath, "utf8");
+
+      // Create the SourceFile AST node
+      const sourceFile = ts.createSourceFile(
+        path.basename(filePath), // Use basename for AST node name
+        sourceText,
+        ts.ScriptTarget.ESNext, // Or your project's target
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:103]: What does this comment `// Or your project's target` mean? It sounds like copypasta :) Please remove it if it is not relevant, or reword it to suggest something actionable.

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+        console.error(`- ${relativePath}:${loc.line}:${loc.character}`);
+      });
+    }
+  });
+
+  if (!duplicatesFound) {
+    log("No duplicate assertion ids found.");
+  }
+  else {
+    process.exit(1);
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:248]: Consider throwing an exception rather than calling process.exit().

File: packages/firestore/scripts/assertion-id-tool.ts [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+  } else if (argv['check']) {
+    handleCheck(allOccurrences);
+  } else {
+    // Default action is List (-L or no flag)
+    handleList(allOccurrences);
+  }
+}
+
+// Run the main function
+main().catch(error => {
```
Comment (dconeybe) [inline packages/firestore/scripts/assertion-id-tool.ts:365]: Can you just let the exception bubble up and crash the program?

**Review by dconeybe — CHANGES_REQUESTED** [review]
Mostly LGTM. Just a few minor comments. Thanks for doing this work! I'll formally "approve" once the comments are addressed/refuted.

File: .changeset/odd-wolves-sit.md (hunk 1 above)
Comment (MarkDuckworth) [reply .changeset/odd-wolves-sit.md]: Updated. Thanks for the suggestion.

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 4 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:59]: That's a good idea. I am going to save this for a future PR. I've hit the timebox for the current effort.

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 3 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:66]: That's clever. Definitely something to consider for a future PR...for the same reason

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 8 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:365]: yeah, implemented

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 7 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:248]: On this one I prefer `process.exit(1)`. The intent is for this to fail the build with a non-zero exit code. Throwing an error could make the failure appear as if there was an issue with the script, rather than an issue with the code it is checking.

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 5 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:168]: Good call. Removed the try catch.

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 6 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:103]: Good catch. Gemini used heavily to move this quickly. I removed this and a few other extraneous comments.

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 2 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:80]: Good call. Logged and re-threw.

File: packages/firestore/scripts/assertion-id-tool.ts (hunk 3 above)
Comment (MarkDuckworth) [reply packages/firestore/scripts/assertion-id-tool.ts:66]: *reason == time boxing

File: packages/firestore/scripts/build-bundle.ts [hunk 9]
```diff
@@ -24,18 +24,20 @@ const typescriptPlugin = require('rollup-plugin-typescript2');
```
Comment (MarkDuckworth) [inline packages/firestore/scripts/build-bundle.ts:24]: Changes in this file are an artifact of running prettier on the scripts folder

File: packages/firestore/scripts/run-tests.ts [hunk 10]
```diff
@@ -20,32 +20,34 @@ import { resolve } from 'path';
 import { spawn } from 'child-process-promise';
```
Comment (MarkDuckworth) [inline packages/firestore/scripts/run-tests.ts:20]: Changes in this file are an artifact of running prettier on the scripts folder
